packages/phone-number-privacy/common/src/interfaces/SignMessageResponse.ts

```diff
@@ -5,4 +5,5 @@ export interface SignMessageResponse {
   error?: string
   performedQueryCount?: number
   totalQuota?: number
+  blockNumber?: number
```

Comment:
I know this isn't really the time or place but I'm curios what other's people's thoughts are on this. I'm not a big fan of having interfaces that have predominantly optional attributes, I think it removes from the benefits of what typescript has to offer.

I'm assuming that the situation here is caused by this response having two possible shapes based on if `success` is true or false, if false, it only has the `error` attribute, if `true` it has all the rest except `error`. 

In this situation I would suggest using static types with literals and a union type. I've setup an example here: https://codesandbox.io/s/jolly-yalow-k026k

The example features two possible implementations:
- `pure.ts` a pure typescript one that does proper typecasting but doesn't do runtime validation if the data is coming from an external source (which is usually the case and here as well)
- `runtypes.ts` which includes the popular [runtypes](https://github.com/pelotom/runtypes) library that allows us to define types that can also be enforced at runtime, the downside being of adding an extra dependency.

I like both options in favour of the current way of doing it and `runtypes` in favour of `pure` because it's an elegant solution for the layer between trust in typescript compiler and trust in your API.

Reply:
Hey @bowd! Thanks for starting this discussion. As someone with a background in strongly typed languages I definitely appreciate the benefits of runtypes.  Part of the reason these are optional is as you mentioned: we need to handle differences in error/success responses. The other two reasons are:
1. Backwards compatibility - signers (9 of them) are operated by us and third parties and we have little control over exactly when they're upgraded. Thus, we must support the case where many of these fields aren't present.
2. Failing gracefully - this service is on the critical path for user onboarding and there are service failure scenarios where we still want to return a partial response to ensure users are not blocked by an outage. Therefore, performedQueryCount, totalQuota, and blockNumber may be unset if there are DB/blockchain connection failures. This also means that signature value may or may not be specified in the case of `success === false`.

So although it won't completely solve the optional attributes problem, I still think it's useful to be explicit at compile time to define the properties we expect to be present and absent. Would like to also here thoughts from @jmrossy and @aslawson on this.

Reply:
i agree on both ends --  i am also quite new to javascript and have a background in strongly typed languages, so it sets off flags to overload an object with optional fields.

i'm not sure how the unions resolve as a javascript object -- depending if the fields are just appended or if the specific fields are nested in another object, it may or may not be doable in a backwards capable manner.

for this use case even if we have the union, many of these fields would be still be optional fields for the success use case so we wouldn't gain much from the refactor (what cody mentioned in 2)